File: backend/models/registration.py
"""Registration mapping utilities."""

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class Registration:
    registration_id: str
    event_id: str
    event_name: str
    full_name: str
    email: str
    phone: str
    status: str
    created_at: str

    def to_api(self) -> dict[str, Any]:
        return {
            "registrationId": self.registration_id, "eventId": self.event_id,
            "eventName": self.event_name, "fullName": self.full_name, "email": self.email,
            "phone": self.phone, "status": self.status, "createdAt": self.created_at,
        }

    @classmethod
    def from_item(cls, item: dict[str, Any]) -> "Registration":
        required_fields = {
            "registration_id", "event_id", "event_name", "full_name",
            "email", "status", "created_at"
        }
        missing = [f for f in required_fields if f not in item or item[f] is None]
        if missing:
            raise ValueError(f"Missing required registration fields: {', '.join(missing)}")
        return cls(
            registration_id=str(item["registration_id"]),
            event_id=str(item["event_id"]),
            event_name=str(item["event_name"]),
            full_name=str(item["full_name"]),
            email=str(item["email"]),
            phone=str(item.get("phone") or ""),
            status=str(item["status"]),
            created_at=str(item["created_at"]),
        )

File: backend/models/registration.js
/**
 * Registration mapping utilities.
 */

class Registration {
  /**
   * @param {Object} props
   * @param {string} props.registrationId
   * @param {string} props.eventId
   * @param {string} props.eventName
   * @param {string} props.fullName
   * @param {string} props.email
   * @param {string} props.phone
   * @param {string} props.status
   * @param {string} props.createdAt
   */
  constructor({ registrationId, eventId, eventName, fullName, email, phone, status, createdAt }) {
    this.registrationId = registrationId;
    this.eventId = eventId;
    this.eventName = eventName;
    this.fullName = fullName;
    this.email = email;
    this.phone = phone;
    this.status = status;
    this.createdAt = createdAt;
  }

  /**
   * Convert to a camelCase API response object.
   *
   * @returns {Object}
   */
  toApi() {
    return {
      registrationId: this.registrationId,
      eventId: this.eventId,
      eventName: this.eventName,
      fullName: this.fullName,
      email: this.email,
      phone: this.phone,
      status: this.status,
      createdAt: this.createdAt,
    };
  }

  /**
   * Create a Registration from a DynamoDB item.
   *
   * @param {Object} item - Raw DynamoDB item.
   * @returns {Registration}
   */
  static fromItem(item) {
    const requiredFields = [
      "registration_id", "event_id", "event_name", "full_name",
      "email", "status", "created_at",
    ];
    const missing = requiredFields.filter(
      (f) => item[f] === undefined || item[f] === null
    );
    if (missing.length > 0) {
      throw new Error(`Missing required registration fields: ${missing.join(", ")}`);
    }
    return new Registration({
      registrationId: String(item.registration_id),
      eventId: String(item.event_id),
      eventName: String(item.event_name),
      fullName: String(item.full_name),
      email: String(item.email),
      phone: String(item.phone || ""),
      status: String(item.status),
      createdAt: String(item.created_at),
    });
  }

  /**
   * Convert to a snake_case DynamoDB item.
   *
   * @returns {Object}
   */
  toDynamoItem() {
    return {
      registration_id: this.registrationId,
      event_id: this.eventId,
      event_name: this.eventName,
      full_name: this.fullName,
      email: this.email,
      phone: this.phone,
      status: this.status,
      created_at: this.createdAt,
    };
  }
}

module.exports = { Registration };
